## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data

df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics

df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [5]:
df.count()

first          54639
last           54639
job            54639
phone          54639
email          54639
client_id      54639
order_id       54639
order_date     54639
order_week     54639
order_year     54639
item_id        54639
category       54639
subcategory    54639
unit_price     54639
unit_cost      54639
unit_weight    54639
qty            54639
line_number    54639
dtype: int64

In [6]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().head(3)
top_categories

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [7]:
# For the category with the most entries,
# which subcategory had the most entries?

top_category = top_categories.index[0]
subcat_most_entries = df[df['category'] == top_category]['subcategory'].value_counts()
subcat_most_entries.head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [8]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
top_clients

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [9]:
# Store the client ids of those top 5 clients in a list.
top_client_ids = top_clients.index.tolist()

top_client_ids

#Onther way-shorter :
#list(top_clients.index)

[33615, 66037, 46820, 38378, 24741]

In [10]:
# How many total units (the qty column) did the
# client with the most entries order order?
most_entries_client = top_client_ids[0]
total_units_most_entries = df[df['client_id'] == most_entries_client]['qty'].sum()
total_units_most_entries

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [11]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df['line_subtotal'] = df['unit_price'] * df['qty']
df[["unit_price", "qty", "line_subtotal"]].head(2)

unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [12]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

#first we have to calculate the total weight for the shippment:
df["total_weight"] = df["unit_weight"] * df["qty"]

#I used Lambda to calculate the price accourding to the weight
df['Shipping_Price'] = df["total_weight"].apply(lambda x: 7 * x if x > 50 else 10 * x)

#this one just to show the columns required as sample provided :
df[["unit_price","unit_weight","qty","total_weight","Shipping_Price"]].head(3)



unit_price  unit_weight  qty  total_weight  Shipping_Price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [13]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%


#Here to find the total Price (line Price)
df["line_price"] = ((df["line_subtotal"] + df["Shipping_Price"]) * 1.0925).round(2)

#Here to show only the columns provided in sample answer- always can use df.head() to show all the df:
df[["line_subtotal", "Shipping_Price", "line_price"]].head(3)


line_subtotal  Shipping_Price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12

In [14]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

df["line_cost"] = ((df["unit_cost"] * df["qty"]) + df["Shipping_Price"]).round(2)

df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  Shipping_Price  line_price  \
0            1      115164.00        787.50         5512.50   131839.08   
1            0         523.95         31.29          312.90      914.26   
2            6         527.28         65.52          458.64     1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [15]:
# Create a column for the profit of
# each line using line cost and line price

df["line_profit"] = (df["line_price"] - df["line_cost"]).round(2)

#Here to show the top 3 rows as the sample provided:
df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  Shipping_Price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [16]:
# Check your work using the totals above
email_list = [2742071,2173913, 6128929] # we creat a list from order IDs Provided

for i in email_list:
    #creat a for loop
    text = df.loc[(df["order_id"]== i) , "line_price"].sum()#.round(2)
    #print it accourding to the text provided:
    print(f"Order {i} total : ${text:,.2f}")
    #can use to print with this line to get exactly same as the text above:
    #print(f"Order ID {i} had a total price of ${text:,.2f}")

Order 2742071 total : $152,811.89
Order 2173913 total : $162,388.71
Order 6128929 total : $923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [17]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

top_five_clients = [33615, 66037, 46820, 38378, 24741]

for client_id in top_five_clients:
     ammount = df.loc[(df["client_id"]== client_id) , "line_price"].sum()
     print(f" {client_id} : ${ammount :.2f}")   

 33615 : $8377308.52
 66037 : $10259514.79
 46820 : $9743794.36
 38378 : $12906550.87
 24741 : $82268892.02


In [18]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

#List of clients and columns needed
top_five_clients = [33615, 66037, 46820, 38378, 24741]
columns_needed = ["qty","Shipping_Price", "line_price", "line_cost", "line_profit"]

summaries = {"client_id" : top_five_clients }

def summarize_clients(client_id,col_name):
    return df.loc[df["client_id"]== client_id , col_name].sum().round(2)
    
for col in columns_needed:
    summaries[col]= [summarize_clients(client_id,col) for client_id in top_five_clients]

    
summary_df = pd.DataFrame(summaries)
    
    
summary_df = summary_df.sort_values(by="line_profit", ascending=False)
    
summary_df


client_id     qty  Shipping_Price   line_price    line_cost  line_profit
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61

In [19]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

#rename the columns:
final_summary =summary_df.rename(columns = { "client_id" :" Client ID",
                                 "qty": "Units",
                                 "Shipping_Price" : "Shipping",
                                 "line_price" : "Total Revenue",
                                 "line_cost" : "Total Cost",
                                 "line_profit" : "Total Profit"
                                 })
#creat function to convert to million and format:
columns_to_convert_m = ["Shipping" , "Total Revenue","Total Cost", "Total Profit"]

def convert_mil(ammount):
    return "${:,.2f} M".format(ammount /1000000)

for col in columns_to_convert_m:
    final_summary[col] = final_summary[col].apply(convert_mil)
    
    
final_summary  

Client ID   Units Shipping Total Revenue Total Cost Total Profit
4       24741  239862  $5.13 M      $82.27 M   $45.69 M     $36.58 M
3       38378   73667  $3.43 M      $12.91 M    $9.63 M      $3.27 M
1       66037   43018  $1.40 M      $10.26 M    $7.00 M      $3.26 M
2       46820   75768  $1.60 M       $9.74 M    $7.01 M      $2.74 M
0       33615   64313  $1.83 M       $8.38 M    $6.18 M      $2.20 M

In [20]:
# Sort the updated data by "Total Profit" form highest to lowest
final_summary = final_summary.sort_values(by="Total Profit" , ascending=False)
final_summary

Client ID   Units Shipping Total Revenue Total Cost Total Profit
4       24741  239862  $5.13 M      $82.27 M   $45.69 M     $36.58 M
3       38378   73667  $3.43 M      $12.91 M    $9.63 M      $3.27 M
1       66037   43018  $1.40 M      $10.26 M    $7.00 M      $3.26 M
2       46820   75768  $1.60 M       $9.74 M    $7.01 M      $2.74 M
0       33615   64313  $1.83 M       $8.38 M    $6.18 M      $2.20 M

______________________________________________________________________________________________________________

### Summary

In summary, the total profit generated by client 24741 amounts to **36.58 million**, surpassing the second-best client by a remarkable 11-fold margin, with a profit of **3.27 million**. Notably, the "cost-profit ratio" for the top client is **80%**, compared to **33.9%** for the second-best client. This underscores the significance of focusing on client 24741, as their purchases yield significantly higher profits.there is a big chance of outliers.

*cost-profit ratio* = *total Profit* / *Total Cost*


In [21]:
final_summary["cost-profit ratio %"] = ((summary_df["line_profit"] / summary_df["line_cost"])*100).round(2)
final_summary

Client ID   Units Shipping Total Revenue Total Cost Total Profit  \
4       24741  239862  $5.13 M      $82.27 M   $45.69 M     $36.58 M   
3       38378   73667  $3.43 M      $12.91 M    $9.63 M      $3.27 M   
1       66037   43018  $1.40 M      $10.26 M    $7.00 M      $3.26 M   
2       46820   75768  $1.60 M       $9.74 M    $7.01 M      $2.74 M   
0       33615   64313  $1.83 M       $8.38 M    $6.18 M      $2.20 M   

   cost-profit ratio %  
4                80.06  
3                33.96  
1                46.47  
2                39.05  
0                35.66

**Highlights:**

- The highest-ranking client, **33615**, purchased more items (top entry in data) and generated a profit of **$2.20 million** with a **35.66%** profit margin.

- Meanwhile, the fifth-ranked client, **24741**, generated a profit of **$36.58 million** with an impressive **80.06%** profit margin.


In [22]:
df.head(2)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   

   line_subtotal  total_weight  Shipping_Price  line_price  line_cost  \
0      115164.00        787.50          5512.5   131839.08   85597.05   
1         523.95         31.29           312.9      914.26     629.79   

   line_profit  
0     46242.03  
1       284.47  

[2 rows x 24 columns]

In [ ]:
#Extra - this is a tool to check if my work was correct or no, we can enter the ID of any client and get the info,
#just want to test and try, there is a lot of things need to be added to this, like checking the type of data and other things



client_id = input("Please Enter Client ID: ")

clients_list = []

if client_id.isdigit():
    clients_list.append(int(client_id))

    for client in clients_list:
        total_revenue = df.loc[df["client_id"] == client,"line_cost"].sum()
        total_profit = df.loc[df["client_id"] == client,"line_profit"].sum()
        print(f"The client ID {client} made a Total Revenue of: $ {total_revenue:,.2f} and total Profit : $ { total_profit:,.2f}")
else:
    print("Please enter a valid Client ID (numerical value).")
    
